### DISCUSSION TITLE
Display warning at comparing enum values with greater/lower than (> / <) operators at extensible enums

### DISCUSSION THREAD

**Discussion by someC0d3r** [StefanMaron/BusinessCentral.LinterCop#325] [New Rule]
We experienced an issue where we check on a enum value if it's bigger than a certain value -> if true then the related document which got the enum value assigned would be classified as (for example) a posted document.

Problem:
The enum is extensible and when somebody extends the enum (in another app) and adds for example values which cannot be classified as a posted document, this would result in a error in the meant comparison above.

Cases:

// OK
if (Rec."Source Document" in [Rec."Source Document"::"Posted S. Invoice", Rec."Source Document"::"Posted S. Shipment"]) then

// NOK -> Warning LC????: You should not use a greater / lower than comparison at enum values
if Rec."Source Document" >= Rec."Source Document"::"Posted S. Invoice" then

**Comment by Arthurvdv** [comment]
This is an easy rule to create I believe. Are there exceptions that I should take into account?

Like for example
```AL
if Rec."Source Document" > 0 then
```

The question is do we need the rule to raise when both sides are a Enum or also when only one side of the comparison is an Enum?

**Reply by someC0d3r** [reply]
Good point - you can also work and compare against the indexes (which in my very honest opinion is "dirty" code but it is still possible :/).

I don't know - maybe consider adding one additional rule?

**First Rule:**
Check if an enum (/option) does get compared to a literal number
Output (Example): "Enum "Sales Document Type" is compared to the number '0'. Please use the identifier syntax instead."
Severity: Info

**Second Rule:**
Check if an enum (/option) does get compared if it is greater or lower than a specific index
Output: "The extensible enum "Sales Document Type" is getting compared by using greater/lower than comparison. Please specify the exact enum value identifiers to prevent that wrong values could be considered."
Severity: Warning

What do you think about this?

**Reply by Arthurvdv** [reply]
Thank for the feedback, and indeed it would be better then to separate this in two rules.

I've looked into both, where I already get a [AL0603](https://learn.microsoft.com/nl-nl/dynamics365/business-central/dev-itpro/developer/diagnostics/diagnostic-al603?wt.mc_id=d365bc_inproduct_alextension) warning on both cases it seems.

```AL
local procedure MyProcedure()
var
    Rec: Record "Warehouse Entry";
begin
    if Rec."Reference Document" >= 0 then
end;
```

```An implicit conversion is being performed from a value of type 'Enum Microsoft.Warehouse.Setup."Whse. Reference Document Type"' to a value of type 'Integer'. This conversion can lead to unexpected runtime issues. This warning will become an error in a future release.```


```AL
local procedure MyProcedure()
var
    Rec: Record "Warehouse Entry";
begin
    if Rec."Reference Document" >= Rec."Reference Document"::"Posted P. Inv." then
end;
```

```An implicit conversion is being performed from a value of type 'Enum Microsoft.Warehouse.Setup."Whse. Reference Document Type"' to a value of type 'Integer'. This conversion can lead to unexpected runtime issues. This warning will become an error in a future release.```

These warning will even be an error in the future 😯

Is there something I'm missing here what the LinterCop should do in addition of the warnings the compiler already shows?

**Reply by someC0d3r** [reply]
Ok - I know this warning but it didn't pop up in my code examples nor in the code of my colleague (which is btw the reason we placed the idea here 😁). I'll ask him and will give you an update here asap.

Thank you so far! 👍

**Reply by Arthurvdv** [reply]
No problem! I also didn't know that there was a rule for this before I've started creating an example.

Could also be that my example is wrong somehow, so looking forward to the feedback of your colleague.

**Reply by Arthurvdv** [reply]
@someC0d3r, do you have any news on this?

**Reply by Arthurvdv** [reply]
@someC0d3r , I’m going to close this discussion. In case we have new insights feel free to re-open this.
